[체스 - 1, 2단계] 도비(김도엽) 미션 제출합니다.

Dobby-Kim opened the pull request:
안녕하세요 파랑!

이번 체스 미션 리뷰 잘 부탁드립니다! ㅎㅎ

지난 블랙잭 미션을 진행하면서 상태 패턴이라는 것을 알게 되었고, 굉장히 인상 깊었던 디자인 패턴이라 페어와 함께 학습해보고 부족한 실력이지만 적용해보았습니다!

상태패턴을 적용해보면서 각 객체가 가져야할 역할과 책임을 명확히 구분해보려고 했습니다!
-특히 모든 행위를 해당 객체에게 메세지를 전달하므로써 판단 및 실행하게 하였습니다!
> ex) 체스판에 존재할 수 있는 Position 인가? -> Position에게 물어보기
> 이동 패턴을 주어졌을 때, 다음 Position은 어디인가? -> Position에게 물어보고 반환받기
> 기물을 `take`할 때: 이동할 `PieceHolder`의 역할을 도착지점의  `PieceHolder` 에게 '이제부터 넌 나야'(이동) 메세지 전달 -> 전달 받은 목적지의 `PieceHolder`는 '너 우리팀이자나 예외 던질게~' or '그래.. 난 여기까지야.. 내 상태를 너로 바꿀게..' 판단 후 실행

## 설계 로직 설명:

1. 현실의 체스와 가장 큰 차이점
- 현실의 체스는 사람이 기물을 직접 '이동' 시키게 됩니다.
- 페어와 저는 기물을 직접 '이동' 시키는 행위가 실제로 객체를 직접 이동시킨다는 행위와 동일하게 취급하지 않게 생각하였습니다.
- 그러므로, 체스 기물을 직접 옮기는 것이 아닌, 기물의 역할을 가지고 있을 수 있는 64개의 `PieceHolder`가 각 체스판에 자리잡고 있고, 그 자리에서 `Role`이라는 상태를 통해 체스 기물의 행위를 행할 수 있게 설계하였습니다.

2. 체스 기물의 이동
- 앞서 설명 드렸듯이 기물 객체의 위치를 옮기는 것이 아닌, 해당 위치의 `PieceHolder`의 상태를 이동 시킬 기물(`Role`)의 상태로 바꾸는 것으로 위치를 이동하는 것처럼 보이게 설계했습니다.
- 해당 위치에 상대 진영의 기물, 혹은 빈칸을 뜻하는 `Square` 상태일 때만 이동을 허용하게 했습니다.
- 이동을 마친 기물 상태는 본래 자리의  `PieceHolder`를 `Square` 상태로 변경하여 결과적으로 이동을 한 것처럼 보이게 하였습니다.

3. 체스판(`ChessBoard`)

- **주요역할: **

    - `ChessBoard`는 `Position`을 Key로 가지며 `PieceHolder`를 Value 값으로 매핑하여 특정 `Position`에 존재하는 `PieceHolder`를 조회하는 역할을 가지고 있습니다.
    - 이동 경로(`Route`)를 시작점의 `PieceHolder`로부터 받아, 그 경로에 존재하는 `PieceHolder`들에게 '네 자리에 뭔가 있니?'라고 물어보고, `Square`들만 있을 때 이동을 허용하게 하는 역할을 가졌습니다.


## 고민하고, 신경썼던 내용

 > 우리가 상태패턴을 제대로 알고, 적재적소에 올바르게 사용하고 있을까?

- 미션 시작부터 제출할 때까지 끊임 없이 고민했던 내용입니다!
- 처음에는 시간이 많이 걸렸었지만, 베이스가 될 도메인 설계를 마친 후에는 부품을 조립하듯이 뚝딱뚝딱 물 흐르듯 코드 작성이 이루어지는 것을 느꼈습니다!
- 그래서 파랑께 상태패턴 적용 부분에서 저희가 보지 못한 부족한 부분에 대해 가감없는 인사이트를 얻고 싶습니다!

Dobby-Kim committed:
docs(README): 1단계 기능 요구 사항 초안 작성

Dobby-Kim committed:
feat(Position): 기물의 위치를 표현하는 객체 생성 기능 구현

Dobby-Kim committed:
feat(Role): Piece에 적용할 상태패턴 구현체 구현

Dobby-Kim committed:
feat(Piece): Piece 클래스 생성

Dobby-Kim committed:
feat(Piece): Piece 클래스 생성

Dobby-Kim committed:
feat(ChessBoard): 체스판 클래스 생성 및 초기화 기능 구현

Dobby-Kim committed:
feat(Position): Position 내의 DTO 정보 전달을 위한 getter 구현

Dobby-Kim committed:
feat(Piece): Piece 내의 DTO 정보 전달을 위한 getter 구현

Dobby-Kim committed:
fix(ChessBoardFactory): Square 상태에 올바른 Color 주입하도록 수정

Dobby-Kim committed:
feat(ChessBoard): ChessBoard DTO 정보 구현을 위한 getter 구현

Dobby-Kim committed:
feat(DTO): Domain을 인자로 받아 View에서 사용할 정보를 추출-전달 하는 DTO Mapper 구현

Dobby-Kim committed:
feat(OutputView): DTO로 전달 받은 정보 출력 기능 구현

Dobby-Kim committed:
feat(Controller): Controller 초기 Chess 실행 메서드 구현

Dobby-Kim committed:
feat(Application): main 함수를 통한 Controller 생성 및 실행 기능 구현

Dobby-Kim committed:
feat(InputView): 사용자 입력 command 처리 기능 구현

Dobby-Kim committed:
refactor(OutputView): 기존의 메서드를 static으로 변경

Dobby-Kim committed:
feat(InputRetryHelper): 사용자 입력 예외 발생 재처리 기능 정적 유틸 클래스 구현

Dobby-Kim committed:
feat(Controller): 사용자 명령에 따른 게임 시작, 종료 기능 구현

Dobby-Kim committed:
docs(README): 기능 요구 사항 진행 업데이트

Dobby-Kim committed:
docs(README): 2단계 기능 요구 사항 추가 및 수정

Dobby-Kim committed:
feat(Direction): Direction Enum 클래스 생성 및 이동방향 검증 기능 구현

- Position 크기 비교 기능 구현

Dobby-Kim committed:
feat(Direction): 해당 방향으로의 이동한 Position 계산 후 반환 기능 구현

Dobby-Kim committed:
refactor(Direction): 클래스명 명시적인 의미로 변경 및 knight 이동 패턴 추가

- Direction -> MovingPattern 클래스명 변경
- Direction 기능 변경에 따른 test 제거

Dobby-Kim committed:
refactor(Position): Position이 이동 가능한 Position 판별 및 반환 기능 구현

Dobby-Kim committed:
feat(Piece.state): 각 Role의 이동 가능한 모든 Position 반환 기능 구현

Dobby-Kim committed:
refactor: Color 멤버 변수 위치 변경

- Piece -> Role로 위치 변경에 따른 코드 변경
- Role: interface -> abstract로 클래스 변경

Dobby-Kim committed:
feat(Role): Pawn의 특수 이동 규칙을 적용한 가능한 Position 반환 기능 구현

- Role의 Color 접근 제어자 protected로 변경

Dobby-Kim committed:
feat(Route): Position들을 관리하며 경로를 의미하는 일급컬렉션 구현

Dobby-Kim committed:
refactor(Role): 시점부터 종점까지의 Route로 포장하여 반환하는 기능 구현

Dobby-Kim committed:
refactor(Position): 패키지 분리 리팩터링

Dobby-Kim committed:
refactor(Piece): 기존 Piece 클래스 역할에 맞는 클래스명으로 변경

- Piece -> PieceHolder

Dobby-Kim committed:
style: 코딩 컨벤션 적용

Dobby-Kim committed:
feat(InputView): 사용자 게임 커맨드 입력 처리 기능 구현

Dobby-Kim committed:
feat(Role): 자신이 자리를 차지하고 있는 기물인지 판단하는 기능 구현

Dobby-Kim committed:
feat(PieceHolder): 상태패턴 기능 적용

Dobby-Kim committed:
feat(Controller): 전체 게임 실행 로직 구현

Dobby-Kim committed:
refactor(OutputView): 게임 실행 시 입력 안내 메세지 변경

Dobby-Kim committed:
refactor(InputView): 예외 발생 메세지 작성 및  매직 넘버 상수화

Dobby-Kim committed:
refactor: 가독성을 위한 패키지, 메서드, 상수, indent 분리

Dobby-Kim committed:
refactor: 메서드명 수정

- getRoute -> findMovingPatternRoute

Dobby-Kim committed:
refactor(Role): 정적 팩토리 메서드로 객체 생성하도록 기능 수정 및 중복 코드 상속

Dobby-Kim committed:
fix(Pawn): Pawn 이동 불가능 에러 수정

Dobby-Kim committed:
feat(ChessBoard): 이동 명령이 차례대로 이루어지고 있는지 검증하는 기능 구현

Dobby-Kim committed:
refactor(Test): test 코드 패키지 이동

Dobby-Kim committed:
refactor(Pawn): Pawn의 첫 이동 특성에 따른 이동가능 루트 계산 기능 리팩터링 적용

Dobby-Kim committed:
refactor(MovingPattern): MovingPattern 클래스명 변경

- MovingPattern -> Direction

summerlunaa commented on README.md:
```diff
@@ -2,6 +2,97 @@
+  - [x] 체스 말(Piece)을 정해진 규칙에 따라 배치한다.
+  - [x] 체스 말(Piece)를 흑/백 구분하여 배치한다.
+  - [x] 체스 말(Piece)을 종류별로 생성한다.
+- [ ] 체스 말(Piece)를 이동시킨다.
```
관리되지 않는 문서는 혼란을 가중시킵니다~ toDoList는 README에서 없애고 개인 메모장에서 작성하는 것도 방법입니다.

summerlunaa commented on src/main/java/controller/Controller.java:
```diff
@@ -0,0 +1,50 @@
+package controller;
+
+import static util.InputRetryHelper.inputRetryHelper;
```
Static Import를 사용하신 이유가 있나요?

Dobby-Kim replied:
제가 작성한 정적 유틸 클래스 특성상 람다식을 이용하여 함수를 실행해야하는데, 가독성이 좋지 않음과 동시에  겹쳐서 혼동될 메서드명이 없어 정적 import를 적용했습니다!

summerlunaa commented on src/main/java/controller/Controller.java:
```diff
@@ -0,0 +1,50 @@
+        Position source = Position.of(file, rank);
+        int toFile = GameCommand.toDestinationFileValue(inputCommand);
+        int toRank = GameCommand.toDestinationRankValue(inputCommand);
+        Position destination = Position.of(toFile, toRank);
```
```suggestion
        Position source = GameCommand.toPosition(inputCommand.get(1));
        Position destination = GameCommand.toPosition(inputCommand.get(2));
```

가능하면 공통으로 사용할 수 있는 범용적인 메서드를 만들어 봅시다. 메서드를 4개에서 1개로 줄일 수 있습니다.

summerlunaa replied:
반영하지 않은 이유가 있을까요?

Dobby-Kim replied:
1. 이동 Commad인 `move src dest `는 그 입력 순서(format)가 명확한 View 단의 입력 요구사항이라고 생각했습니다!

따라서 Controller가 Command에 존재하는 index를 통해 get을 사용하여 파라미터로 넘겨주면 view의 요구사항을 알고 침범하기에 명확한 역할 분리를 위해 코드를 분리했습니다.

2. GameCommand 클래스는 입력된 Command를 처리 및 변환하여 Controller에게 그 정보를 넘겨주는 역할을 가지고 있습니다.
그래서 도메인에서 가장 주요 로직으로 사용되는 `Position` 객체를  `GameCommand`가 직접 생성하지 않도록 분리가 필요하다고 생각했습니다.

위의 각 두 이유에 의해 Source와 Destination을 각각 구하는 메서드 2개가 필요했고, 각 File과 Rank를 구하는 메서드 2개가 필요해서 4개의 메서드로 나눴습니다!

summerlunaa commented on src/main/java/model/chessboard/ChessBoardFactory.java:
```diff
@@ -0,0 +1,100 @@
+
+public class ChessBoardFactory {
+
+    private static final int FILE_A = 1;
```
[[Effective java] item 34. int 상수 대신 열거 타입을 사용하라](https://newwisdom.tistory.com/17)

Dobby-Kim committed:
refactor(Role): 상속 단계 추가 구현

- Role -> *ShiftRole -> 기존 기물들
- 상속 관계 변화에 따른 메서드 접근 제어자 변경
- TestCode 변경
- Route에 Direction field 추가

summerlunaa commented on src/main/java/model/piece/state/Bishop.java:
```diff
@@ -0,0 +1,22 @@
+import model.direction.MovingPattern;
+import model.piece.Color;
+
+public final class Bishop extends Role {
```
final 클래스 👍

Dobby-Kim committed:
refactor(Role): Direction list field를 ShiftPattern enum으로 변경

Dobby-Kim committed:
refactor(Role): 메서드명 통일

summerlunaa commented on src/main/java/view/dto/InfoMapper.java:
```diff
@@ -0,0 +1,24 @@
+        return chessBoard.getChessBoard()
+                .entrySet()
+                .stream()
+                .map(entry -> new PieceInfo(
+                        entry.getKey().file(),
+                        entry.getKey().rank(),
+                        entry.getValue().getRole().getClass().getSimpleName(),
+                        entry.getValue().color().name())
```
- 모든 원시 값과 문자열을 포장한다
- 3개 이상의 인스턴스 변수를 가진 클래스를 쓰지 않는다.
- 한 줄에 점을 하나만 찍는다.
- 게터/세터/프로퍼티를 쓰지 않는다.

몇 개를 어긴 코드인가요? 프로그래밍 요구 사항을 모든 상황에서 지킬 필요는 없지만 명확한 이유가 없다면 따르는 게 좋겠죠~! Enum을 잘 활용해보면 좋을 것 같습니다.

Dobby-Kim committed:
refactor(Role): Color 접근제어자 변경

Dobby-Kim committed:
refactor(ShiftPattern): 미사용 field인 bound 제거

summerlunaa commented on src/main/java/model/piece/state/Role.java:
```diff
@@ -0,0 +1,59 @@
+import model.position.Position;
+import model.position.Route;
+
+public abstract class Role {
```
상태패턴을 적용했다고 해서 패키지 이름이 state여야 한다거나 클래스명이 달라져야 하는 것은 아닙니다. 체스의 기물이라는 의미가 더 잘 드러나는 네이밍을 사용하면 더 좋을 것 같아요.

summerlunaa commented on src/main/java/model/piece/PieceHolder.java:
```diff
@@ -0,0 +1,39 @@
+import model.position.Position;
+import model.position.Route;
+
+public class PieceHolder {
```
오호 상태 패턴을 적용하려 하셨군요! 현재 구조 자체는 좋지만 상태패턴이라고 보기는 어려울 것 같습니다 😂
상태패턴은 `객체의 상태에 따라 객체의 행위를 변경할 수 있도록 하는 것`이 핵심입니다. 체스에서 각 칸은 어떤 기물이 존재하느냐에 따라 행위가 달라진다고 보기는 어렵습니다. `changeRoleTo` 메서드에서 보듯이 다음에 위치할 기물을 외부에서 계산하여 넣어주는 방식이죠.

보통 상태 다이어그램으로 나타낼 수 있는 로직을 구현할 때 상태 패턴을 적용해 볼 수 있습니다. 체스 미션에서 상태패턴을 적용한다면 기물의 움직임 보다는 체스 게임의 흐름을 구현하는 데에 적용할 수 있습니다.

게임 준비 -> 시작 -> white turn -> black turn -> (반복) -> finishied 

이런 식으로요. 근데 뭐 어떤 패턴인지가 중요한 건 아닙니다. 상속을 통해 잘 구현해주신 것 같아요~

summerlunaa reviewed (changes requested):
도비 안녕하세요~ 체스 미션의 리뷰어 파랑입니다 🐳

레벨 1의 마지막 미션까지 오셨네요! 특히나 이번 미션은 쉽지 않았을 것 같은데 시간 맞춰서 잘 구현해주셨습니다ㅎㅎ

첫 리뷰인데다 양도 많은 만큼 리뷰를 많이 드리지는 않았습니다. 짧게 자주 티키타카 해보면 좋을 것 같아요. 필요하시면 언제든 편하게 DM 주시구요~ 감사합니다 ☺️

summerlunaa commented on src/main/java/model/piece/state/Role.java:
```diff
@@ -0,0 +1,53 @@
+        return possibleRoutes;
+    }
+
+    protected abstract Route findRouteByDirection(Direction direction, Position source);
```
리뷰하는 사이에 리팩토링 해주셨네요! 
각 클래스별로 기능이 다르다면 이미 구현된 메서드를 오버라이딩하는 것보단 이렇게 추상 메서드를 사용하는 것이 더 안전합니다 👍  MultiShiftRole과 SingleShiftRole로 잘 나눠주신 것 같아요.

Dobby-Kim committed:
feat(Color): 자신의 다음 차례의 색깔을 반환하는 기능 구현

Dobby-Kim committed:
refactor(PieceHolder & Role): PieceHolder가 스스로 자신의 상태를 변환하도록 기능 수정

Dobby-Kim committed:
refactor(Route): 경로에 목적지까지 포함하는 subList 반환하도록 수정

Dobby-Kim committed:
refactor(Pawn): Pawn의 특수 움직임 기능 구현

Dobby-Kim committed:
docs(README): 기능 요구 사항 진행 상황 업데이트

Dobby-Kim committed:
refactor(model.state): 패키지명 변경

- state -> role

Dobby-Kim committed:
refactor(Controller): 중복 코드 메서드 분리 적용

Dobby-Kim committed:
refactor(Role): 부모 상속 클래스 변경

- SingleShiftRole -> Role

Dobby-Kim committed:
refactor(GameCommand): 입력받은 file, rank 변환 enum class 적용

Dobby-Kim committed:
refactor(Position): 매직넘버 상수화 적용

Dobby-Kim committed:
refactor(File,Rank Info): file, rank enum util로 패키지 변경

Dobby-Kim committed:
refactor(ChessBoardFactory): 매직 넘버 상수 제거 후 Enum Converter 클래스 이용

Dobby-Kim committed:
refactor(Pawn): 도착지에 같은 색깔의 기물인지 확인하는 기능 추가

Dobby-Kim committed:
test(PawnTest): Pawn 이동 기능 테스트코드 추가

Dobby-Kim committed:
refactor(Role): 멤버 변수 접근 제어자 변경

- Color: private -> protected

Dobby-Kim committed:
refactor(ChessBoard): 메서드 순서 변경

Dobby-Kim committed:
feat(Direction): 반대방향 반환하는 기능 구현

Dobby-Kim committed:
feat(ShiftPattern): 특정 방향을 가지고 있는지 판단하는 기능 구현

Dobby-Kim committed:
feat(Role): 해당 PieceHolder의 Role이 King인지 판별하는 기능 구현

Dobby-Kim committed:
style(ChessBoard): 코딩 컨벤션 적용

Dobby-Kim committed:
refactor(Color): Enum 클래스 비교 로직 중 equals 제거

Dobby-Kim committed:
refactor(Position): Position 클래스 static final int 제거 및 메서드 예외 처리 수정

Dobby-Kim committed:
style(Role): 코딩 컨벤션 적용

Dobby-Kim committed:
feat: ChessBoard에서 턴이 끝난 후 check인지 확인하는 기능 구현

Dobby-Kim committed:
feat(PieceHolder): 메서드명 변경

Dobby-Kim committed:
refactor(QueenTest): 불필요 코드 제거

Dobby-Kim committed:
refactor(Controller): 메서드 기능 분리

- 출발점, 도착점 구하는 기능 분리

Dobby-Kim committed:
refactor(ChessBoard): 상태 패턴 기본 디자인 적용

Dobby-Kim committed:
feat(Route): 특정 Route의 반대 방향으로 향하는 Route를 구하는 기능 구현

Dobby-Kim committed:
refactor(Pawn): 메서드 명확성을 위한 네이밍 변경

Dobby-Kim committed:
refactor(PieceHolder): 이동 가능한 루트인지 판별하는 메서드명 수정

Dobby-Kim committed:
refactor(DefaultState): 메서드 접근제어자 변경

- protected nextState() -> public nextState()

Dobby-Kim committed:
refactor(Controller): 메서드 위치 변경 및 마지막 이동 결과 출력

Dobby-Kim committed:
feat(DefaultState): 상태별 상세 기능 구현

Dobby-Kim committed:
feat(DefaultState): 상태별 상세 기능 구현

Dobby-Kim committed:
Merge branch 'step1' of https://github.com/Dobby-Kim/java-chess into step1

Dobby-Kim commented:
안녕하세요 파랑!

리뷰요청이 조금 늦었습니다!
체크 메이트까지 어떻게 하면 완벽하게 구현하고 싶은 욕심이 생겨서 이것 저것 시도해보다가 좀 늦은 것 같네요😅😅

실질적인 기능 구현을 하다보니, `PieceHolder` <-> `DefaultState` 간에 다른 기물의 Position 과 Role 정보 조회 및 전달에 비효율이 있는 것을 확인했습니다!
다음 리픽토링 때에는 이 부분에 집중해보고자 합니다.


큰 변화로는 `DefaultState`라는 상태가 체스 보드에 적용되었습니다. 각 상태는 현재 차례의 행동을 실행한 후, 발생할 수 있는 상태들을 반환하는 형식으로 구성되어있습니다.


체크와 체크메이트를 구현하면서 기존 코드는 많이 바뀌지 않았습니다. 다만, DefaultState에 존재하는 코드가 좀 복잡해진 것 같습니다🥲🥲
이 복잡성은 앞서 말씀드린 `PieceHolder` <-> `DefaultState` 간의 비효율 때문에 발생하는 것 같아 수정해보고자 합니다!

summerlunaa commented on src/main/java/model/direction/ShiftPattern.java:
```diff
@@ -0,0 +1,46 @@
+
+    public boolean contains(Direction direction) {
+        return directions.contains(direction);
+    }
```
사용하지 않는 메서드네요 👀

summerlunaa commented on src/main/java/view/dto/InfoMapper.java:
```diff
@@ -0,0 +1,31 @@
+        return chessBoard.getChessBoard()
+                .entrySet()
+                .stream()
+                .map(entry -> new PieceInfo(
+                        entry.getKey()
+                                .file(),
+                        entry.getKey()
+                                .rank(),
+                        entry.getValue()
+                                .getRole()
+                                .getClass()
+                                .getSimpleName(),
+                        entry.getValue()
+                                .getColor()
+                                .name())
```
단순히 줄바꿈만 하라는 의미는 아니었는데요, 도메인 객체가 View에 존재하지 않도록 구현하신 것 같은데 이유가 무엇인가요? 
role은 왜 클래스명을 사용하고 있나요? 클래스명이 변경되면 출력에도 영향이 미쳐야할까요?

Dobby-Kim replied:
체스라는 강력한 도메인 규칙에 종속되어있는 모든 입출력-메인로직은 그 기물의 이름에 변화가 없을 것이라고 생각하여 클래스명을 사용하였습니다.

다만, InitialPawn과 같은 클래스가 생길 수도 있을 것이라는 생각이 들어 DTO의 정보를 출력 로직으로 변경시켜주는 `ChessSymbol` 클래스에서 클래스 명과 정확하게 일치하는지 판단하는 equals가 아닌, contains를 이용하는 것으로 변경하였습니다!
도메인의 클래스명이라도 체스의 명명 규칙 때문에 기본적으로 본래의 이름을 포함하고 있을 것이라고 생각해서 이렇게 적용해봤습니다!

summerlunaa replied:
기물에 대한 출력이 변경될 수도 있고 클래스명이 변경될 수도 있습니다. 클래스명을 출력에 사용하는 건 여러모로 위험한 방법입니다. 다음 리뷰 요청때 수정해주세요~

summerlunaa commented on src/main/java/model/position/Position.java:
```diff
@@ -0,0 +1,79 @@
+    }
+
+    public static Position of(int file, int rank) {
+        if (A.getValue() <= file && file <= H.getValue() && ONE.getValue() <= rank && rank <= EIGHT.getValue()) {
```
비슷한 코드가 아래에서도 반복되네요. 값을 꺼내오지 않고 검증할 방법을 고민해보면 좋겠습니다.

summerlunaa commented on src/main/java/util/FileConverter.java:
```diff
@@ -0,0 +1,33 @@
+
+import java.util.Arrays;
+
+public enum FileConverter {
```
File, Rank라는 객체를 사용하지 않고 Converter를 둔 이유가 잘 이해되지 않습니다. File, Rank 값이 여기저기서 원시값으로 돌아다니고 있어요.
추가로 도비는 util 패키지에 어떤 클래스들이 들어간다고 생각하나요?

Dobby-Kim replied:
File과 Rank 또한 체스라는 강력한 도메인 규칙에 종속되어 있기에 입출력, 도메인 로직 전반에 모두 통용적으로 사용될 것이라고 생각했습니다.

그렇기에 코드 전역적으로 사용되는 클래스로 판단되어 static util 패키지에 존재하며 유틸 클래스로서의 역할을 맡아야된다고 생각했습니다!

그리고 많이 고민을 한 부분이 있었는데, `Position` 클래스 자체가 file과 rank를 감싸는 래핑 클래스라고 결론을 내렸었습니다!

만일 File과 Rank 래핑 클래스를 각각 따로 만들어 Position의 file로 둔다면 Position 내부적인 가공으로 변경할 수 있는 File, Rank 값들을 가공 및 변경하기 위한 불필요한 코드 중복과 Depth가 깊어지며 getter를 반복적으로 사용해야될 것이라고 생각했습니다!

많이 고민했던 부분이라 이 래핑과 관련된 부분에서 파랑의 인사이트를 얻고 싶습니다!

summerlunaa replied:
단순히 프로젝트에 전역적으로 사용된다고 해서 util 클래스인가요? 어떤 클래스가 util 클래스인지 다시 고민해보시면 좋을 것 같아요.
File과 Rank는 비즈니스 로직이 담긴 도메인입니다. 출력이 분리되어야 한다면 클래스 자체를 분리하는 것이 맞지 이도저도 아닌 Converter 클래스는 사용성만 떨어뜨리지 않나 싶습니다 ^^;
File, Rank 클래스를 만드는 경우 값의 연산이 필요하면 메서드를 통해 처리할 수 있습니다. getter를 사용하지 않고 값을 효율적으로 사용할 수 있는 방법을 고민해보세용~

Dobby-Kim committed:
docs(README): 기능 요구 사항 목록 업데이트

summerlunaa commented on src/main/java/controller/Controller.java:
```diff
@@ -0,0 +1,60 @@
+        return gameCommand;
+    }
+
+    private void controlChessBoard(ChessBoard chessBoard, List<String> inputCommand) {
```
체스보드를 어떻게 컨트롤 한다는 걸까요? 메서드명을 더 적절하게 수정해봅시다~

summerlunaa commented on src/main/java/model/chessboard/ChessBoardFactory.java:
```diff
@@ -0,0 +1,88 @@
+
+    private static void initWhiteRookBishopKnight() {
+        List<Role> whiteTypePieces = generateColoredPieces(WHITE);
+        for (int file = A.getValue(); file <= C.getValue(); file++) {
```
FILE과 RANK를 사용하는 모든 곳에서 static 임포트를 사용했네요. 고민해볼 부분이 많은 포인트인 것 같아요.

- 흔하게 사용되는 A, B, C 알파벳과 ONE, TWO 숫자를 클래스명을 명시하지 않고 사용했을 때 FILE과 RANK의 의미가 확실하게 전달되는가 (맥락을 확인해야만 의미를 알 수 있는 건 아닌지)
- static import를 사용하지 않았을 때 `FileConverter.A`, `RankConverter.ONE` 처럼 사용될텐데 클래스명이 적절한지
- 사용되는 모든 곳에서 `getValue`를 함께 사용하는 것이 적절한지

고민해보시고 수정 부탁드립니다.

summerlunaa replied:
요기도 도비의 의견을 남겨주세요오~~

summerlunaa commented on src/main/java/model/chessboard/ChessBoardFactory.java:
```diff
@@ -0,0 +1,88 @@
+        }
+    }
+
+    private static void initBlackRookBishopKnight() {
```
initWhiteRookBishopKnight과 initBlackRookBishopKnight을 초기화하는 방식이 이해하기 어렵습니다. initKingQueen 메서드처럼 단순하게 변경하면 훨씬 이해하기 좋을 것 같아요 👍

summerlunaa commented on src/main/java/model/chessboard/state/CurrentTurn.java:
```diff
@@ -0,0 +1,98 @@
+        }
+    }
+
+    private void checkTurn(PieceHolder selectedPieceHolderColor) {
```
메서드가 사용되는 맥락을 고려하지 않는 적절한 파라미터명으로 변경해볼까요?

summerlunaa commented on src/main/java/model/chessboard/state/CurrentTurn.java:
```diff
@@ -0,0 +1,98 @@
+        checkTurn(sourcePieceHolder);
+        Route route = sourcePieceHolder.findRoute(source, destination);
+        runMove(sourcePieceHolder, route);
+        if (isCheckedBy(currentColor)) {
+            return new Checked(chessBoard, currentColor.opponent(), route.reverseRouteTowardSource());
```
<img width="80" alt="image" src="https://github.com/woowacourse/java-chess/assets/77425729/0232f52e-2ad8-4754-9d0d-f15d3fdf8303">

이 상태도 체그메이트인가요? K은 피할 수 없지만 P가 n을 잡으면 충분히 게임을 이어나갈 수 있을 것 같아요. 체크메이트 로직을 다시 한번 살펴봐주세요..!

Dobby-Kim replied:
맞습니다ㅜㅜ 안그래도 치명적인 실수를 발견해서 수정을 시도하고 있었습니다!

summerlunaa replied:
요거 수정 안 된 것 같은데 확인해보시고 다음에 반영해주세욥

summerlunaa commented on src/main/java/model/chessboard/state/DefaultState.java:
```diff
@@ -0,0 +1,28 @@
+import model.piece.PieceHolder;
+import model.position.Position;
+
+public abstract class DefaultState {
```
상태패턴을 적용하셨네요ㅎㅎ 말씀해주신 것처럼 내부 로직이 너무 복잡해서 이해하기 쉽지는 않은 것 같습니다. 로직과 네이밍을 더 간결하게 리팩토링하면 더 이해하고 관리하기 편해지겠네요 👍 

직접 적용해보니 도비는 어떤 걸 느꼈나요? 어떤 부분이 더 좋아졌다 느끼는지, 단점은 없었는지 궁금합니다.

Dobby-Kim replied:
확실히 각 상태일 때 해야하는 행동이 명확하게 분리되어 한 클래스 내에서의 코드 양과 분기문이 줄어드는 것을 느꼈습니다!
또한, 공통적으로 해야하는 행동들에 대한 코드를 공유하니 추가적인 조건이 발동된 상태에서만 코드를 추가하는 것이 편하게 느껴졌습니다ㅎㅎ

다만, 상속을 통해서 여러 클래스가 하나의 코드를 공유하다보니 코드 작성을 할 때 상위 클래스를 작성할 때 만들었던 메서드들을 계속 적절하게 포괄할 수 있는 메서드명으로 바꿔야하거나, 접근제어자를 변경해야하는 등의 불편함이 있었던 것 같습니다.

summerlunaa reviewed (changes requested):
도비 안녕하세요~

리팩토링하느라 고생하셨습니다! 리뷰 남겨드렸으니 확인 부탁드립니다~

Dobby-Kim committed:
refactor(File, Rank): 클래스명 변경, 검증 기능 추가 구현

- 클래스명 및 메서드명 변경사항 전역적 적용

Dobby-Kim committed:
refactor(Position): 메서드명 오타 수정 및 전역 적용

Dobby-Kim committed:
fix(PieceInfo): 버그 수정

Dobby-Kim committed:
refactor(ShiftPattern): 미사용 메서드 제거

summerlunaa commented on README.md:
다음 단계에서 아래 내용에 대해 고민해보고 의견 나눠보면 좋겠습니다.

- 도비는 왜 뷰와 도메인을 명확하게 나누고 있나요?
- 뷰와 도메인을 명확하게 나누기 위해서, 혹은 패턴을 적용해서 클래스가 많아지고 로직이 복잡해진 부분들도 있는데요, 로직이 복잡해졌을 때의 단점은 무엇이 있을까요?
- 원시값을 포장하지 않고 사용했을 때의 단점은 무엇이 있을까요?
- 프로젝트에서 전역적으로 사용되는 클래스는 모두 util인가요?

추가로 리뷰를 반영하거나, 혹은 반영하지 않은 이유를 남겨주시면 더 좋은 대화를 나눌 수 있을 것 같아요! 참고 부탁드립니다 🙇‍♀️

summerlunaa reviewed (approved):
도비 안녕하세요~ 

1단계는 여기서 머지하겠습니다. 리뷰에도 코멘트 다시 남겨드린 부분이 있으니 확인해보시고 다음 PR에서 또 얘기 나눠보면 좋겠습니다! 수고하셨습니다.
